eskipfile/file.go

```diff
@@ -16,45 +16,105 @@
 
+It supports runtime updates of routes by watching changes to the file.
+
 (See the DataClient interface in the skipper/routing package and the eskip
 format in the skipper/eskip package.)
 */
 package eskipfile
 
 import (
-	"github.com/zalando/skipper/eskip"
 	"io/ioutil"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/rjeczalik/notify"
+	"github.com/zalando/skipper/eskip"
 )
 
 // A Client contains the route definitions from an eskip file.
-type Client struct{ routes []*eskip.Route }
+type Client struct {
+	path     string
+	routes   []*eskip.Route
+	fsevents chan notify.EventInfo
+}
 
-// Opens an eskip file and parses it, returning a DataClient implementation.
+// Open reads an eskip file and parses it, returning a DataClient implementation.
 // If reading or parsing the file fails, returns an error.
-func Open(path string) (*Client, error) {
-	content, err := ioutil.ReadFile(path)
+func Open(p string) (*Client, error) {
+	r, err := readRoutesFromFile(p)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{routes: r, path: p}, nil
+}
+
+// Watch behaves like Open but it starts watching the eskip file for changes.
+// This enabled runtime updates of routes when using this DataClient.
+// The filesystem watcher created by this call should be properly Closed using this Client's Close function.
+func Watch(p string) (*Client, error) {
+	c, err := Open(p)
 	if err != nil {
 		return nil, err
 	}
+	c.fsevents = make(chan notify.EventInfo, 1)
+	// We only care about Write events and Rename for "atomic" saves from some editors
+	if err := notify.Watch(c.path, c.fsevents, notify.Write, notify.Rename); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Close does a clean shutdown of the Client. It is not needed unless the Client was created using the
+// Watch function, since it creates a filesystem watcher
+func (c *Client) Close() {
+	if c.fsevents != nil {
+		notify.Stop(c.fsevents)
+	}
+}
 
-	routes, err := eskip.Parse(string(content))
+func readRoutesFromFile(path string) ([]*eskip.Route, error) {
+	content, err := ioutil.ReadFile(path)
 	if err != nil {
 		return nil, err
 	}
 
-	return &Client{routes}, nil
+	return eskip.Parse(string(content))
 }
 
-func (c Client) LoadAndParseAll() (routeInfos []*eskip.RouteInfo, err error) {
+func (c *Client) LoadAndParseAll() (routeInfos []*eskip.RouteInfo, err error) {
 	for _, route := range c.routes {
 		routeInfos = append(routeInfos, &eskip.RouteInfo{Route: *route})
 	}
 	return
 }
 
 // Returns the parsed route definitions found in the file.
-func (c Client) LoadAll() ([]*eskip.Route, error) { return c.routes, nil }
+func (c *Client) LoadAll() ([]*eskip.Route, error) { return c.routes, nil }
+
+// LoadUpdate is No-Op if the Client was created using Open.
+// When created using Watch, it monitors the filesystem for changes on the eskip file and enables runtime updates
+// of the routes
+func (c *Client) LoadUpdate() ([]*eskip.Route, []string, error) {
```

Comment:
As far as I understood, this part of the DataClient interface is used to report the IDs of deleted routes. Maybe @aryszka can confirm.